src/Microsoft.Extensions.Options/Options.cs

```diff
@@ -1,6 +1,71 @@
-﻿namespace Microsoft.Extensions.Options
+﻿using System;
+
+namespace Microsoft.Extensions.Options
 {
     /// <summary>
+    /// Implementation of IOptionsFactory.
+    /// </summary>
+    /// <typeparam name="TOptions">The type of options being requested.</typeparam>
+    public class OptionsService<TOptions> : IOptions<TOptions>, IOptionsSnapshot<TOptions> where TOptions : class, new()
+    {
+        private readonly IOptionsCache<TOptions> _cache;
+        private readonly IOptionsFactory<TOptions> _factory;
+
+        /// <summary>
+        /// Initializes a new instance with the specified options configurations.
+        /// </summary>
+        /// <param name="cache">The cache to use.</param>
+        /// <param name="factory">The factory to use to create options.</param>
+        public OptionsService(IOptionsCache<TOptions> cache, IOptionsFactory<TOptions> factory)
+        {
+            _cache = cache;
+            _factory = factory;
+        }
+
+        public TOptions Value
+        {
+            get
+            {
+                return Get("");
+            }
+        }
+
+        public virtual void Add(string name, TOptions options)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+            if (!_cache.TryAdd(name, options))
+            {
+                throw new InvalidOperationException("An option named {name} already exists.");
+            }
+        }
+
+        public virtual TOptions Get(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            return _cache.GetOrAdd(name, () => _factory.Create(name));
```

Comment:
>_cache.GetOrAdd(name, () => _factory.Create(name)); [](start = 19, length = 51)

Initial ideas for async. Separate out the creation and population of options into a new service--something like IOptionsSource. It depends on IOptionsFactory and IOptionsCache. For the normal, non-async, lazy route, OptionsService just depends on this and does what is happening here. But for those who want greater control, the IOptionsSource can be resolved directly and then a Create or Initialize method called that executes the factory, if needed, and stores the result in the cache. On this interface there can be an async overload that will call an async method on the factory. The idea being that the experience and patterns stay the same for normal users, but if you need async. then you can force async resolve explicitly before requesting the options.